README.md
<p align="center">
    <img src="https://vmstan.com/content/images/2021/02/gs-logo.svg" width="40%" alt="Gravity Sync">
</p>

What is better than a [Pi-hole](https://github.com/pi-hole/pi-hole) blocking ads via DNS on your network? That's right, **two** Pi-hole blocking ads on your network!

- [Seriously. Why two Pi-hole?](https://github.com/vmstan/gravity-sync/wiki/Frequent-Questions#why-do-i-need-more-than-one-pi-hole)

But if you have more than one Pi-hole in your network you'll want a simple way to keep the list configurations and local DNS settings identical between the two. That's Gravity Sync. With proper preparation it should only take a few minutes to install. Ideally you set up Gravity Sync and forget about it -- and in the long term, it would be awesome if the Pi-hole team made this entire script unnecessary.

# Features

Gravity Sync replicates the `gravity.db` database, which includes:

- Blocklist settings with status and comments.
- Domain whitelist and blacklist along with status with comments.
- Custom RegEx whitelist and blacklists.
- Clients and groups along with any list assignments.

Gravity Sync can also (optionally) replicate FTLDNS/DNSMASQ configuration files, including:

- Local DNS (A Records) which are stored in a separate `custom.list` file within the `/etc/pihole` directory.
- CNAME Records which are stored in a separate `05-pihole-custom-cname.conf` file in the `/etc/dnsmasq.d` directory.

### Limitations

Gravity Sync will **not**:

- Overwrite device specific Pi-hole settings specific to the local network configuration.
- Change the Pi-hole admin/API passwords, nor does not leverage these at all.
- Modify the individual Pi-hole's upstream DNS resolvers.
- Sync DHCP settings or monitor device leases.
- Merge long term data, query logs, or statistics.

# Setup Steps

1. [Review System Requirements](https://github.com/vmstan/gravity-sync/wiki/System-Requirements)
2. [Prepare Your Pi-hole](https://github.com/vmstan/gravity-sync/wiki/Installing#primary-pi-hole)
3. [Install Gravity Sync](https://github.com/vmstan/gravity-sync/wiki/Installing#secondary-pi-hole) (or [Upgrade](https://github.com/vmstan/gravity-sync/wiki/Updating))
4. [Configure Gravity Sync](https://github.com/vmstan/gravity-sync/wiki/Installing#configuration)
5. [Execute Gravity Sync](https://github.com/vmstan/gravity-sync/wiki/Installing#execution)
6. [Automate Gravity Sync](https://github.com/vmstan/gravity-sync/wiki/Installing#automation)
7. [Profit](https://memory-alpha.fandom.com/wiki/Rules_of_Acquisition)

# Disclaimer

Gravity Sync is not developed by or affiliated with the Pi-hole project. This is project an unofficial, community effort, that seeks to implement replication (which is currently not a part of the core Pi-hole product) in a way that provides stability and value to Pi-hole users. The code has been tested across multiple user environments but there always is an element of risk involved with running any arbitrary software you find on the Internet.

Pi-hole is and the Pi-hole logo are [registered trademarks](https://pi-hole.net/trademark-rules-and-brand-guidelines/) of Pi-hole LLC.

# Additional Documentation

Please refer to the [Wiki](https://github.com/vmstan/gravity-sync/wiki) for more information:

- [Frequently Asked Questions](https://github.com/vmstan/gravity-sync/wiki/Frequent-Questions)
- [Advanced Installation Options](https://github.com/vmstan/gravity-sync/wiki/Under-The-Covers)
- [Changelog](https://github.com/vmstan/gravity-sync/blob/master/CHANGELOG.md)


CHANGELOG.md
# The Changelog

## 3.3

### The Podman Release

For this release, "beta" support for [Podman](https://podman.io) based container deployments has been added. (Thanks to work by [@mfschumann](https://github.com/vmstan/gravity-sync/pull/138)) This is marked as beta because at the moment, Gravity Sync may not be regularly tested with this container engine and so regular support may be limited. Use of the official Pi-hole container image is still required.

This release also removes automated nightly backups as a function of the automation script. You can still execute a manual `./gravity-sync.sh backup` anytime. Automating this function was made redundant by the primary synchronization functions backing up the database files prior to execution in later versions, resulting in multiple backups or backing up unchanged data. Existing users will continue to backup each night unless you run `./gravity-sync.sh automate` again to configure a new schedule, at which time the existing backup job will be deleted.

Additionally, this release focuses on making some of the prompts and messages in Gravity Sync easier to understand. Starting with the initial install and configuration wizard. As the script has grown, added features, and become more complex... more options are necessary during install and it wasn't always clear what to do. This release adds some clarification to various components and will change over time.

#### Configuration Changes

- The first time you execute Gravity Sync after upgrading to 3.3, your `gravity-sync.conf` file will be moved into a `settings` folder in the same directory.
- The first time you execute Gravity Sync after upgrading to 3.3, your existing `gravity-sync.md5`, `gravity-sync.log`, and `gravity-sync.cron` files will be moved into a `logs` folder in the same directory.
- The default number days for which backups are retained has been reduced from 7 to 3.

#### Bug Fixes

- Docker Swarm use for the Pi-hole container should be supported by a change in the execution command.

#### 3.3.1

- Fixes issue with inverted Advanced Install Options selection [#163](https://github.com/vmstan/gravity-sync/issues/163)

## 3.2

### The Alias Release

This release now fully supports Pi-hole 5.2, specifically the CNAME replication features that were added. Because the location of these settings is in a new directory that was not previously monitored by Gravity Sync, you will need to opt-in to replication by updating your configuration file to enable support for replicating this.

- New setups can be prompted to enable this during configuration using the Advanced Configuration options.
- Existing installs wishing to make sure of this feature, should either re-run the `./gravity-sync config` command, or manually edit the `gravity-sync.conf` to set `INCLUDE_CNAME='1'`.
- Before you enable `INCLUDE_CNAME` make sure you've created at least one entry from within the primary Pi-hole interface, otherwise Gravity Sync will have nothing to replicate as the files will not yet exist.
- You cannot enable `INCLUDE_CNAME` if you've also enabled `SKIP_CUSTOM` as the CNAME function is dependent on Local DNS records. You can, however, only replicate the Local DNS Records if you do not intend to leverage CNAME records.
- Existing installs using Docker, or otherwise using a non-standard location for dnsmasq configuration files (default is `/etc/dnsmasq.d`) will also need to manually specify the location of these files.
- See the [Hidden Figures](https://github.com/vmstan/gravity-sync/wiki/Hidden-Figures) document for more details.

#### More Syncing Coming

Even before the Pi-hole team added the CNAME feature and implemented in such a way that the `/etc/dnsmasq.d` would need to be seen by Gravity Sync, I have had a desire to replicate additional custom files here for my own selfish needs. More people asked for a similar function, and now that it's required to be built into the core script, it's easier to include these additional files.

Not implemented in 3.2.0, but coming within this release, Gravity Sync will be configured to monitor a custom file of your creation in this folder (default is `08-gs-lan.conf`) which can contain additional configuration options for DNSMASQ.

An example would be setting different caching options for Pi-hole, or specifying the lookup targets for additional networks. Similar requirements as above for the CNAME syncing must be met for existing installs to leverage this functionality.

#### 3.2.1

- Changes application of permissions for Docker instances to UID:GID instead of names. (#99/#128)
- Adds `./gravity-sync.sh info` function to help with troubleshooting installation/configuration settings.

#### 3.2.2

The `./gravity-sync.sh restore` process completely revamped:

- You can now choose to ignore any of the three elements during restore.
- The prompts are clearer and more consistent with Gravity Sync script styling.
- Importantly, lack of backup files in an element will not cause the restoration to fail.

#### 3.2.3

- Cleanup of the error message screen when an invalid command is run (ex: `./gravity-sync.sh wtf`)

#### 3.2.4

- Changes script startup and shutdown text format.

#### 3.2.5

- Correct error where Docker based installs would fail to restart if Docker exec commands required sudo privileges.
- Correct error where setup script would prompt twice for Local/Remote DNSMASQ directories when using Docker.

#### 3.2.6

- Adds old backup removal tasks into push/pull/sync/compare functions.

## 3.1

### The Container Release

The premise of this release was to focus on adding better support for Docker container instances of Pi-hole. This release also changes a lot of things about the requirements that Gravity Sync has always had, which were not running as the root user, and requiring that the script be run from the user's home directory. Those two restrictions are gone.


gravity-sync.sh
#!/bin/bash
SCRIPT_START=$SECONDS

# GRAVITY SYNC BY VMSTAN #####################
PROGRAM='Gravity Sync'
VERSION='3.3.3'


# For documentation or downloading updates visit https://github.com/vmstan/gravity-sync
# Requires Pi-Hole 5.x or higher already be installed, for help visit https://pi-hole.net

# REQUIRED SETTINGS ##########################

# Run './gravity-sync.sh config' to get started, it will customize the script for your environment
# You should not to change the values of any variables here here to customize your install
# Add replacement variables to gravity-sync.conf, which will overwrite these defaults.

# STANDARD VARIABLES #########################

# Installation Types
PH_IN_TYPE='default'				# Pi-hole install type, `default`, `docker`, or `podman` (local)
RH_IN_TYPE='default'				# Pi-hole install type, `default`, `docker`, or `podman` (remote)

# Pi-hole Folder/File Customization
PIHOLE_DIR='/etc/pihole' 			# default Pi-hole data directory (local)
RIHOLE_DIR='/etc/pihole'			# default Pi-hole data directory (remote)
DNSMAQ_DIR='/etc/dnsmasq.d'         # default DNSMASQ data directory (local)
RNSMAQ_DIR='/etc/dnsmasq.d'         # default DNSMASQ data directory (remote)
PIHOLE_BIN='/usr/local/bin/pihole' 	# default Pi-hole binary directory (local)
RIHOLE_BIN='/usr/local/bin/pihole' 	# default Pi-hole binary directory (remote)
DOCKER_BIN='/usr/bin/docker'		# default Docker binary directory (local)
ROCKER_BIN='/usr/bin/docker'		# default Docker binary directory (remote)
PODMAN_BIN='/usr/bin/podman'        # default Podman binary directory (local)
RODMAN_BIN='/usr/bin/podman'        # default Podman binary directory (remote)
FILE_OWNER='pihole:pihole'			# default Pi-hole file owner and group (local)
RILE_OWNER='pihole:pihole'			# default Pi-hole file owner and group (remote)
DOCKER_CON='pihole'					# default Pi-hole container name (local)
ROCKER_CON='pihole'					# default Pi-hole container name (remote)
CONTAIMAGE='pihole/pihole'          # official Pi-hole container image

GRAVITY_FI='gravity.db' 			        # default Pi-hole database file
CUSTOM_DNS='custom.list'			        # default Pi-hole local DNS lookups
CNAME_CONF='05-pihole-custom-cname.conf'    # default DNSMASQ CNAME alias file
GSLAN_CONF='08-gs-lan.conf'                 # default DNSMASQ GS managed file

# Interaction Customization
VERIFY_PASS='0'						# replace in gravity-sync.conf to overwrite
SKIP_CUSTOM='0'						# replace in gravity-sync.conf to overwrite
INCLUDE_CNAME='0'					# replace in gravity-sync.conf to overwrite
DATE_OUTPUT='0'						# replace in gravity-sync.conf to overwrite
PING_AVOID='0'						# replace in gravity-sync.conf to overwrite
ROOT_CHECK_AVOID='0'				# replace in gravity-sync.conf to overwrite

# Backup Customization
BACKUP_RETAIN='3'					# replace in gravity-sync.conf to overwrite

# SSH Customization
SSH_PORT='22' 						# default SSH port
SSH_PKIF='.ssh/id_rsa'				# default local SSH key

# GS Folder/File Locations
GS_FILEPATH=$(realpath $0)			# auto determined - do not change!
LOCAL_FOLDR=$(dirname $GS_FILEPATH) # auto determined - do not change!
CONFIG_FILE='gravity-sync.conf' 	# must exist with primary host/user configured
GS_FILENAME='gravity-sync.sh'		# must exist because it's this script
BACKUP_FOLD='backup' 				# must exist as subdirectory in LOCAL_FOLDR
LOG_PATH="${LOCAL_FOLDR}/logs"		# replace in gravity-sync.conf to overwrite
SYNCING_LOG='gravity-sync.log' 		# replace in gravity-sync.conf to overwrite
CRONJOB_LOG='gravity-sync.cron' 	# replace in gravity-sync.conf to overwrite
HISTORY_MD5='gravity-sync.md5'		# replace in gravity-sync.conf to overwrite

# OS Settings
BASH_PATH='/bin/bash'				# default OS bash path

##############################################
### NEVER CHANGE ANYTHING BELOW THIS LINE! ###
##############################################

# Import UI Fields
source ${LOCAL_FOLDR}/includes/gs-ui.sh

# Import Color/Message Includes
source ${LOCAL_FOLDR}/includes/gs-colors.sh

# FUNCTION DEFINITIONS #######################

# Core Functions
source ${LOCAL_FOLDR}/includes/gs-core.sh

# Gravity Replication Functions
source ${LOCAL_FOLDR}/includes/gs-compare.sh
source ${LOCAL_FOLDR}/includes/gs-pull.sh
source ${LOCAL_FOLDR}/includes/gs-push.sh
source ${LOCAL_FOLDR}/includes/gs-smart.sh
source ${LOCAL_FOLDR}/includes/gs-restore.sh
source ${LOCAL_FOLDR}/includes/gs-backup.sh

# Hashing & SSH Functions
source ${LOCAL_FOLDR}/includes/gs-hashing.sh
source ${LOCAL_FOLDR}/includes/gs-ssh.sh

# Logging Functions
source ${LOCAL_FOLDR}/includes/gs-logging.sh

# Validation Functions
source ${LOCAL_FOLDR}/includes/gs-validate.sh
source ${LOCAL_FOLDR}/includes/gs-intent.sh
source ${LOCAL_FOLDR}/includes/gs-root.sh

# Configuration Management
source ${LOCAL_FOLDR}/includes/gs-config.sh
source ${LOCAL_FOLDR}/includes/gs-update.sh
source ${LOCAL_FOLDR}/includes/gs-automate.sh
source ${LOCAL_FOLDR}/includes/gs-purge.sh

# Exit Codes
source ${LOCAL_FOLDR}/includes/gs-exit.sh

# SCRIPT EXECUTION ###########################

case $# in
    0)
        start_gs
    task_smart ;;
    1)
        case $1 in
            smart|sync)
                start_gs
                task_smart ;;
            pull)
                start_gs
                task_pull ;;
            push)
                start_gs
                task_push ;;
            restore)
                start_gs
                task_restore ;;
            version)
                start_gs_noconfig
                task_version ;;
            update|upgrade)
                start_gs_noconfig
                task_update ;;
            dev|devmode|development|develop)
                start_gs_noconfig
                task_devmode ;;
            logs|log)
                start_gs
                task_logs ;;
            compare)
                start_gs
                task_compare ;;
            cron)
                start_gs
                task_cron ;;
            config|configure)
                start_gs_noconfig
                task_configure ;;
            auto|automate)
                start_gs
                task_automate ;;
            backup)
                start_gs
                task_backup ;;
            purge)
                start_gs
                task_purge ;;
            sudo)
                start_gs
                task_sudo ;;
            info)
                start_gs
                task_info ;;
            *)
                start_gs
                task_invalid ;;
        esac
    ;;
    
    2)
        case $1 in
            auto|automate)
                start_gs
                task_automate ;;
        esac
    ;;
    
    3)
        case $1 in
            auto|automate)
                start_gs
                task_automate $2 ;;
        esac
    ;;
    
    *)
        start_gs
        task_invalid ;;
esac

# END OF SCRIPT ##############################